realm/src/main/java/io/realm/RealmObject.java

```diff
@@ -124,4 +136,137 @@ protected static Realm getRealm(RealmObject obj) {
+        this.clazz = clazz;
+    }
+
+    /**
+     * Determine if the current RealmObject is obtained synchronously or asynchronously (from
+     * a worker thread). Synchronously RealmObjects are by definition blocking hence this method
+     * will always return {@code true} for them.
+     *
+     * @return {@code true} if the query has completed & the data is available {@code false} if the
+     * query is in progress.
+     */
+    public boolean isLoaded() {
+        realm.checkIfValid();
+        return pendingQuery == null || isCompleted;
+    }
+
+    /**
+     * make an asynchronous query blocking.
+     *
+     * @return {@code true} if it successfully completed the query, {@code false} otherwise.
+     */
+    public boolean load() {
+        // doesn't guarantee to import correctly the result (because the user may have advanced)
+        // in this case the Realm#handler will be responsible of retrying
+        realm.checkIfValid();
+        if (pendingQuery != null && !pendingQuery.isDone() && !isCompleted) {
+            return onCompleted();
+        }
+        return false;
+    }
+
+    /**
+     * called to import the handover row pointer & notify listeners.
+     *
+     * @return {@code true} if it successfully completed the query, {@code false} otherwise.
+     */
+    boolean onCompleted() {
+        realm.checkIfValid();
+        try {
+            Long handoverResult = pendingQuery.get();// make the query blocking
+            // this may fail with BadVersionException if the caller and/or the worker thread
+            // are not in sync. REALM_COMPLETED_ASYNC_FIND_FIRST will be fired by the worker thread
+            // this is should handle more complex use cases like retry, ignore etc
+            onCompleted(handoverResult);
+        } catch (Exception e) {
+            RealmLog.d(e.getMessage());
+            return false;
+        }
+        return true;
+    }
+
+    void onCompleted(Long handoverRowPointer) {
+        realm.checkIfValid();
+        if (!isCompleted) {
+            isCompleted = true;
+            long nativeRowPointer = TableQuery.nativeImportHandoverRowIntoSharedGroup(handoverRowPointer, realm.sharedGroup.getNativePointer());
+            Table table = realm.getTable(clazz);
+            this.row = table.getUncheckedRowByPointer(nativeRowPointer);
+            notifyChangeListeners();
+        }// else: already loaded query no need to import again the pointer
+    }
+
+    /**
+     * add a change listener to this RealmResults
+     *
+     * @param listener the change listener to be notified
+     */
+    public void addChangeListener(RealmChangeListener listener) {
+        if (listener == null) {
+            throw new IllegalArgumentException("Listener should not be null");
+        }
+        realm.checkIfValid();
+
+        if (!listeners.contains(listener)) {
+            listeners.add(listener);
+        }
+    }
+
+    /**
+     * remove a previously registered listener
```

Comment:
In general all JavaDoc should read as proper sentences. Start with capital letter, end with .
Will not comment on these further below.